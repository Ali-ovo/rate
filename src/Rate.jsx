import ReactDOM from 'react-dom';
import React, { PropTypes } from 'react';
import classNames from 'classnames';
import { getOffsetLeft } from './util';
import Star from './Star';

function noop() {
}

const Rate = React.createClass({
  propTypes: {
    disabled: PropTypes.bool,
    value: PropTypes.number,
    defaultValue: PropTypes.number,
    count: PropTypes.number,
    allowHalf: PropTypes.bool,
    style: PropTypes.object,
    prefixCls: PropTypes.string,
    onChange: PropTypes.func,
    onHoverChange: PropTypes.func,
    className: PropTypes.string,
    character: PropTypes.node,
  },

  getDefaultProps() {
    return {
      defaultValue: 0,
      count: 5,
      allowHalf: false,
      style: {},
      prefixCls: 'rc-rate',
      onChange: noop,
      character: 'â˜…',
      onHoverChange: noop,
    };
  },

  getInitialState() {
    let value = this.props.value;
    if (value === undefined) {
      value = this.props.defaultValue;
    }
    return {
      value,
    };
  },

  componentWillReceiveProps(nextProps) {
    if ('value' in nextProps) {
      let value = nextProps.value;
      if (value === undefined) {
        value = nextProps.defaultValue;
      }
      this.setState({
        value,
      });
    }
  },

  onHover(event, index) {
    const hoverValue = this.getStarValue(index, event.pageX);
    this.setState({
      hoverValue,
    });
    this.props.onHoverChange(hoverValue);
  },

  onMouseLeave() {
    this.setState({
      hoverValue: undefined,
    });
    this.props.onHoverChange(undefined);
  },

  onClick(event, index) {
    const value = this.getStarValue(index, event.pageX);
    if (!('value' in this.props)) {
      this.setState({
        value,
      });
    }
    this.onMouseLeave();
    this.props.onChange(value);
  },

  getStarDOM(index) {
    return ReactDOM.findDOMNode(this.refs[`star_${index}`]);
  },

  getStarValue(index, x) {
    let value = index + 1;
    if (this.props.allowHalf) {
      const leftEdge = getOffsetLeft(this.getStarDOM(0));
      const width = getOffsetLeft(this.getStarDOM(1)) - leftEdge;
      if ((x - leftEdge - width * index) < width / 2) {
        value -= 0.5;
      }
    }
    return value;
  },

  render() {
    const { count, allowHalf, style, prefixCls, disabled, className, character } = this.props;
    const { value, hoverValue } = this.state;
    const stars = [];
    const disabledClass = disabled ? `${prefixCls}-disabled` : '';
    for (let index = 0; index < count; index++) {
      stars.push(
        <Star
          ref={`star_${index}`}
          index={index}
          disabled={disabled}
          prefixCls={`${prefixCls}-star`}
          allowHalf={allowHalf}
          value={hoverValue === undefined ? value : hoverValue}
          onClick={this.onClick}
          onHover={this.onHover}
          key={index}
          character={character}
        />
      );
    }
    return (
      <ul
        className={classNames(prefixCls, disabledClass, className)}
        style={style}
        onMouseLeave={disabled ? null : this.onMouseLeave}
      >
        {stars}
      </ul>
    );
  },
});

export default Rate;
