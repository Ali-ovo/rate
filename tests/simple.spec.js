import React from 'react';
import { render, mount } from 'enzyme';
import KeyCode from 'rc-util/lib/KeyCode';
import Rate from '../src';

describe('rate', () => {
  describe('full', () => {
    it('render works', () => {
      const wrapper = render(
        <Rate count={3} value={1} className="custom" />
      );
      expect(wrapper).toMatchSnapshot();
    });

    it('support focus and blur', () => {
      const wrapper = mount(
        <Rate count={3} value={2} />
      );
      wrapper.simulate('focus');
      expect(
        wrapper.find('li').at(1).hasClass('rc-rate-star-focused')
      ).toBe(true);

      wrapper.simulate('blur');
      expect(
        wrapper.find('li').at(1).hasClass('rc-rate-star-focused')
      ).toBe(false);
    });

    it('support keyboard', () => {
      const handleChange = jest.fn();
      const wrapper = mount(
        <Rate count={3} value={1} onChange={handleChange} />
      );
      wrapper.simulate('keyDown', { keyCode: KeyCode.LEFT });
      expect(handleChange).toBeCalledWith(0);
      handleChange.mockReset();
      wrapper.simulate('keyDown', { keyCode: KeyCode.RIGHT });
      expect(handleChange).toBeCalledWith(2);
    });
  });

  describe('allowHalf', () => {
    it('render works', () => {
      const wrapper = render(
        <Rate count={3} value={1.5} allowHalf className="custom" />
      );
      expect(wrapper).toMatchSnapshot();
    });

    it('support focus and blur', () => {
      const wrapper = mount(
        <Rate count={3} value={1.5} allowHalf />
      );
      wrapper.simulate('focus');
      expect(
        wrapper.find('li').at(1).hasClass('rc-rate-star-focused')
      ).toBe(true);

      wrapper.simulate('blur');
      expect(
        wrapper.find('li').at(1).hasClass('rc-rate-star-focused')
      ).toBe(false);
    });

    it('support keyboard', () => {
      const handleChange = jest.fn();
      const wrapper = mount(
        <Rate count={3} value={1.5} allowHalf onChange={handleChange} />
      );
      wrapper.simulate('keyDown', { keyCode: KeyCode.LEFT });
      expect(handleChange).toBeCalledWith(1);
      handleChange.mockReset();
      wrapper.simulate('keyDown', { keyCode: KeyCode.RIGHT });
      expect(handleChange).toBeCalledWith(2);
    });
  });
});
