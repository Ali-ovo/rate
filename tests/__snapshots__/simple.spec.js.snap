// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rate allowHalf render works 1`] = `
<ul
  class="rc-rate custom"
  tabindex="0"
>
  <li
    class="rc-rate-star rc-rate-star-full"
  >
    <div
      class="rc-rate-star-first"
    >
      ★
    </div>
    <div
      class="rc-rate-star-second"
    >
      ★
    </div>
  </li>
  <li
    class="rc-rate-star rc-rate-star-half rc-rate-star-active"
  >
    <div
      class="rc-rate-star-first"
    >
      ★
    </div>
    <div
      class="rc-rate-star-second"
    >
      ★
    </div>
  </li>
  <li
    class="rc-rate-star rc-rate-star-zero"
  >
    <div
      class="rc-rate-star-first"
    >
      ★
    </div>
    <div
      class="rc-rate-star-second"
    >
      ★
    </div>
  </li>
</ul>
`;

exports[`rate full render works 1`] = `
<ul
  class="rc-rate custom"
  tabindex="0"
>
  <li
    class="rc-rate-star rc-rate-star-full"
  >
    <div
      class="rc-rate-star-first"
    >
      ★
    </div>
    <div
      class="rc-rate-star-second"
    >
      ★
    </div>
  </li>
  <li
    class="rc-rate-star rc-rate-star-zero"
  >
    <div
      class="rc-rate-star-first"
    >
      ★
    </div>
    <div
      class="rc-rate-star-second"
    >
      ★
    </div>
  </li>
  <li
    class="rc-rate-star rc-rate-star-zero"
  >
    <div
      class="rc-rate-star-first"
    >
      ★
    </div>
    <div
      class="rc-rate-star-second"
    >
      ★
    </div>
  </li>
</ul>
`;
